### PR TITLE
SQLAlchemy for Django developers

### PR DESCRIPTION
**Pull request by jerivas**
## Steps to test/reproduce
https://deploy-preview-431--oddleventy.netlify.app/2023/10/20/sqlalchemy-for-django-developers/

### CODE REVIEW

**Comment by jerivas** [comment]
[I want an article about SQL Alchemy](https://trello.com/c/oyHFAR3N/232-i-want-an-article-about-sql-alchemy)

File: content/blog/2023/sqlalchemy-for-django-developers.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+need to use a `join` to bring in the `User` model and then filter on the `name`
+attribute.
+
+One clear advantage of SQLAlchemy is that you can instantiate model instances
+and relate them before they are saved, and then save them all at once:
+
+```python
+spongebob = User(
+    name="spongebob",
+    fullname="Spongebob Squarepants",
```
Comment (jgerigmeyer) [inline content/blog/2023/sqlalchemy-for-django-developers.md:287]:
```suggestion
    fullname="SpongeBob SquarePants",
```

File: content/blog/2023/sqlalchemy-for-django-developers.md [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+4. How to Use FastAPI Dependency Injection Everywhere (coming soon)
+
+[FastAPI Path Operations for Django Developers]: ./fastapi-path-operations-for-django-developers
+
+## About SQLAlchemy
+
+Both Django's ORM (Object Relational Mapper) and SQLAlchemy are libraries for
+interacting with relational databases. Django's ORM is tightly coupled with the
+Django framework, while SQLAlchemy is a standalone library that can be used with
+any Python application. I had heard about it for many years but never had the
```
Comment (jgerigmeyer) [inline content/blog/2023/sqlalchemy-for-django-developers.md:39]:
```suggestion
any Python application. I heard about it many years ago, but never had the
```

File: content/blog/2023/sqlalchemy-for-django-developers.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+patrick = User(name="patrick", fullname="Patrick Star")
+session.add_all([spongebob, sandy, patrick])
+session.commit()
+```
+
+In Django you would need to save all the users first, then save the addresses
+with the `user` attribute set to the corresponding user. Here we can nest them
+naturally and SQLAlchemy will take care of the rest.
+
+## Migrations are not Built-in
```
Comment (jgerigmeyer) [inline content/blog/2023/sqlalchemy-for-django-developers.md:307]:
```suggestion
## Migrations Are Not Built-In
```

**Review by jgerigmeyer — COMMENTED** [review]
👏

File: content/blog/2023/sqlalchemy-for-django-developers.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+We won't go into details here, but the basic substitutions are:
+
+- `./manage.py makemigrations` becomes `alembic revision --autogenerate`
+- `./manage.py migrate` becomes `alembic upgrade head`
+- `./manage.py migrate app <migration number>` becomes `alembic upgrade
+  <migration hash>` if going forward, or `alembic downgrade <migration
+  hash>` if going back
+
+To date I don't know how to manage migrations in databases that have models
+defined by third party packages.
```
Comment (jamesnw) [inline content/blog/2023/sqlalchemy-for-django-developers.md:326]: Would it be worth giving an example of when you might run into this?

**Review by jamesnw — APPROVED** [review]
LGTM!

File: content/blog/2023/sqlalchemy-for-django-developers.md [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+These two models are similar in that once they get information out of the
+database, table columns are accessible as attributes of the model instances.
+However, they differ in that SQLAlchemy relies more heavily on type hints to
+define the model. In my opinion this results in a more robust developer
+experience, and the `Mapped` annotation enables some nice tricks we will see
+later.
+
+Another similarity is that modifying the attributes of a model instance will
+modify the corresponding database record when the object is "saved" (which is
+done differently in SQLAlchemy as we will see later).
```
[the remark this answers is not in the document]
Comment (jerivas) [reply content/blog/2023/sqlalchemy-for-django-developers.md]: Thanks, it is confusing! I reworked this section in ed2dcbfb674fb0820bcd55a53c9b73038f330438

File: content/blog/2023/sqlalchemy-for-django-developers.md (hunk 4 above)
Comment (jerivas) [reply content/blog/2023/sqlalchemy-for-django-developers.md]: Added in ed2dcbfb674fb0820bcd55a53c9b73038f330438

File: content/blog/2023/fastapi-path-operations-for-django-developers.md [hunk 6]
```diff
@@ -24,10 +24,12 @@ on our recent experience converting an internal project from Django to FastAPI.
 [FastAPI]: https://fastapi.tiangolo.com/
 
 1. FastAPI Path Operations for Django Developers (this article)
-2. SQLAlchemy for Django Developers (coming soon)
+2. [SQLAlchemy for Django Developers]
 3. Testing a FastAPI Application (coming soon)
 4. How To Use FastAPI Dependency Injection Everywhere (coming soon)
 
+[SQLAlchemy for Django Developers]: /2023/10/20/sqlalchemy-for-django-developers/
```
Comment (jerivas) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:31]: Is there a more generic way to get this links? Maybe linking to a post just by name and letting 11ty add the right date

File: content/blog/2023/sqlalchemy-for-django-developers.md [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+These two models are similar in that once they get information out of the
+database, table columns are accessible as attributes of the model instances such
+as `user.name` and `user.fullname`. In the same way, modifying the attributes of
+a model instance like `user.name = "Jane"` will modify the corresponding
+database record when the object is "saved" (which is done differently in
+SQLAlchemy as we will see later).
+
+In contrast, these models differ in that SQLAlchemy relies more heavily on type
+hints to define the model. In my opinion this results in a more robust developer
+experience, and the `Mapped` annotation enables some nice tricks we will see
```
Comment (jamesnw) [inline content/blog/2023/sqlalchemy-for-django-developers.md:94]: Would it be worth calling out later on what Mapped is enabling? I saw it was in use, but didn't see easily what the nice tricks were.

**Review by jamesnw — APPROVED** [review]
🚀

File: content/blog/2023/sqlalchemy-for-django-developers.md [hunk 8]
```diff
@@ -1,7 +1,7 @@
 ---
 title: SQLAlchemy for Django Developers
 author: ed
-date: 2023-10-24
+date: 2023-10-20
```
Comment (jgerigmeyer) [inline content/blog/2023/sqlalchemy-for-django-developers.md:4]: Is there a benefit to spacing these out just a bit, and publishing this article next week?

**Review by jgerigmeyer — APPROVED** [review]

File: content/blog/2023/sqlalchemy-for-django-developers.md (hunk 7 above)
Comment (jerivas) [reply content/blog/2023/sqlalchemy-for-django-developers.md]: Thanks, I tried to be more explicit in 835a72433fa9486a1d16f1a870ee0c1aa1e50a2c

File: content/blog/2023/sqlalchemy-for-django-developers.md (hunk 8 above)
Comment (jerivas) [reply content/blog/2023/sqlalchemy-for-django-developers.md:4]: Pushed it to Monday on 822f7c0d1909275b8ce25cccf7851f82a2eb72d0

File: content/blog/2023/fastapi-path-operations-for-django-developers.md (hunk 6 above)
Comment (jgerigmeyer) [reply content/blog/2023/fastapi-path-operations-for-django-developers.md:31]: Not that I'm aware of...

Comment (jgerigmeyer) [inline content/blog/2023/fastapi-path-operations-for-django-developers.md:31]:
```suggestion
[SQLAlchemy for Django Developers]: /2023/10/23/sqlalchemy-for-django-developers/
```

**Review by jgerigmeyer — APPROVED** [review]

**Pull request merged**